File: src/main/java/voidchess/figures/RochadeFigure.java
package voidchess.figures;

import voidchess.helper.Move;
import voidchess.helper.Position;

/**
 * @author stephan
 */
abstract public class RochadeFigure extends Figure {
    protected int stepsTaken;
    final private Position initialPos;

    public RochadeFigure(boolean isWhite, Position startPosition, FigureType type) {
        this(isWhite, startPosition, 0, type);
    }

    public RochadeFigure(boolean isWhite, Position startPosition, int stepsTaken, FigureType type) {
        super(isWhite, startPosition, type);
        initialPos = startPosition;
        this.stepsTaken = stepsTaken;
    }

    public Position getInitialPosition() {
        return initialPos;
    }

    public boolean canParticipateInRochade() {
        return stepsTaken == 0;
    }

    public void undoMove(Position oldPosition) {
        stepsTaken--;
        super.undoMove(oldPosition);
    }

    public void figureMoved(Move move) {
        if (getPosition().equalsPosition(move.getFrom())) stepsTaken++;
        super.figureMoved(move);
    }

    public String toString() {
        StringBuilder buffer = new StringBuilder(20);
        buffer.append(super.toString());
        buffer.append("-");
        buffer.append(stepsTaken);
        return buffer.toString();
    }

}

File: src/main/kotlin/voidchess/figures/RochadeFigure.kt
package voidchess.figures

import voidchess.helper.Move
import voidchess.helper.Position

/**
 * @author stephan
 */
abstract class RochadeFigure(isWhite: Boolean, val initialPosition: Position, protected var stepsTaken: Int, type: FigureType) : Figure(isWhite, initialPosition, type) {

    constructor(isWhite: Boolean, startPosition: Position, type: FigureType) : this(isWhite, startPosition, 0, type)

    override fun undoMove(oldPosition: Position) {
        stepsTaken--
        super.undoMove(oldPosition)
    }

    override fun figureMoved(move: Move) {
        if (position.equalsPosition(move.from)) stepsTaken++
        super.figureMoved(move)
    }

    override fun canParticipateInRochade() = stepsTaken == 0
    override fun toString() = "${super.toString()}-$stepsTaken"
}
